substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/typestate/DefaultAnalysisPolicy.java

```diff
@@ -337,7 +337,7 @@ public TypeState doUnion(PointsToAnalysis bb, MultiTypeState s1, MultiTypeState
          * No need for a deep equality check (which would need to iterate the arrays), since the
          * speculation logic below is doing that anyway.
          */
-        if (s1.typesBitSet() == s2.typesBitSet()) {
+        if (s1.bitSetEquals(s2)) {
```

Comment:
Switching from `==` to `equals()`  invalidates the speculation.

Reply:
Ah, good point. I will make something that does the shallow check without materializing a `BitSet`.